File: runtime/cpp/cyana-ast/RuntimeAutomataAstApplication.cpp
//
// Created by tian wei jun on 2022/11/23 0023.
//

#include "RuntimeAutomataAstApplication.h"
#include <list>

RuntimeAutomataAstApplication::RuntimeAutomataAstApplication() : persistentAutomataAstApplication(0) {
}

RuntimeAutomataAstApplication::~RuntimeAutomataAstApplication() {
  delete persistentAutomataAstApplication;
  persistentAutomataAstApplication = 0;
}

void RuntimeAutomataAstApplication::setContext(const std::string *automataFilePath) {
  persistentAutomataAstApplication = new PersistentAutomataAstApplication(automataFilePath);
}

const Ast *RuntimeAutomataAstApplication::buildAst(const std::string *sourceCodeFilePath) {
  return persistentAutomataAstApplication->buildAst(sourceCodeFilePath);
}

File: runtime/java/cyana-ast-runtime/src/cyana/ast/runtime/RuntimeAutomataAstApplication.java
package cyana.ast.runtime;

import java.io.InputStream;

/**
 * RuntimeAutomataAstApplication.
 *
 * @author tian wei jun
 */
public class RuntimeAutomataAstApplication {
  PersistentAutomataAstApplication persistentAutomataAstApplication;

  public Ast buildAst(String sourceFilePath) {
    return persistentAutomataAstApplication.buildAst(sourceFilePath);
  }

  public Ast buildAst(InputStream sourceInputStream) {
    Ast ast = persistentAutomataAstApplication.buildAst(sourceInputStream);
    try {
      if (null != sourceInputStream) {
        sourceInputStream.close();
      }
    } catch (Exception e) {
      throw new CyanaAstRuntimeException(e);
    }
    return ast;
  }

  public void setContext(InputStream automataInputStream) {
    // automataInputStream 会被下层模块关闭
    persistentAutomataAstApplication = new PersistentAutomataAstApplication(automataInputStream);
  }

  public void setContext(String automataFilePath) {
    persistentAutomataAstApplication = new PersistentAutomataAstApplication(automataFilePath);
  }

  public void displayGraphicalViewOfAst(Ast ast) {
    new cyana.ast.runtime.AstGuiOutputer().output(ast);
  }
}
